File: solver/transitions.py
"""Combinatorial building blocks for the Yahtzee DP (§3.1).

Enumerates the two canonical multiset spaces and the reroll transition table:

- ``ROLLS`` — the 252 distinct 5-dice hands (multisets over 6 faces).
- ``KEEPS`` — the 462 distinct keep sets (multisets of size 0..5).
- ``T`` — a 462 x 252 matrix; ``T[k, r]`` is the probability that keeping set ``k`` and
  rerolling the remaining dice yields hand ``r``. Each row is a distribution (sums to 1).
- ``roll_prob`` — the initial-throw distribution over the 252 hands (= the empty-keep row).
- ``subkeep_flat`` / ``subkeep_starts`` — a flattened index of, for each hand, the keep
  sets that are sub-multisets of it, laid out for ``np.maximum.reduceat`` so the within-turn
  "best keep" reduction is a single vectorized op.

A hand/keep is a length-6 tuple of face counts (face ``f`` -> index ``f-1``). The canonical
ordering is: enumerate multisets by ``itertools.combinations_with_replacement`` on faces,
keeps grouped by ascending size. This ordering is the shared spec the TS port must match.
"""

from __future__ import annotations

import math
from itertools import combinations_with_replacement

import numpy as np

NUM_FACES = 6
HAND_SIZE = 5


def _counts_from_faces(faces: tuple[int, ...]) -> tuple[int, ...]:
    counts = [0] * NUM_FACES
    for f in faces:
        counts[f] += 1
    return tuple(counts)


def _enumerate_multisets(size: int) -> list[tuple[int, ...]]:
    return [
        _counts_from_faces(combo)
        for combo in combinations_with_replacement(range(NUM_FACES), size)
    ]


def _multinomial(total: int, counts: tuple[int, ...]) -> int:
    result = math.factorial(total)
    for n in counts:
        result //= math.factorial(n)
    return result


ROLLS: list[tuple[int, ...]] = _enumerate_multisets(HAND_SIZE)
KEEPS: list[tuple[int, ...]] = [
    ms for size in range(HAND_SIZE + 1) for ms in _enumerate_multisets(size)
]

NUM_ROLLS = len(ROLLS)
NUM_KEEPS = len(KEEPS)
assert NUM_ROLLS == 252, NUM_ROLLS
assert NUM_KEEPS == 462, NUM_KEEPS

ROLL_INDEX = {counts: i for i, counts in enumerate(ROLLS)}
KEEP_INDEX = {counts: i for i, counts in enumerate(KEEPS)}
EMPTY_KEEP = KEEP_INDEX[(0,) * NUM_FACES]


def _build_transition() -> np.ndarray:
    T = np.zeros((NUM_KEEPS, NUM_ROLLS), dtype=np.float64)
    for ki, keep in enumerate(KEEPS):
        rerolls = HAND_SIZE - sum(keep)
        denom = NUM_FACES**rerolls
        for ri, roll in enumerate(ROLLS):
            delta = tuple(roll[f] - keep[f] for f in range(NUM_FACES))
            if any(d < 0 for d in delta):
                continue  # keep is not a sub-multiset of this roll
            T[ki, ri] = _multinomial(rerolls, delta) / denom
    return T


T = _build_transition()
roll_prob = T[EMPTY_KEEP].copy()


def _build_subkeeps() -> tuple[np.ndarray, np.ndarray]:
    flat: list[int] = []
    starts = np.empty(NUM_ROLLS, dtype=np.int64)
    for ri, roll in enumerate(ROLLS):
        starts[ri] = len(flat)
        for ki, keep in enumerate(KEEPS):
            if all(keep[f] <= roll[f] for f in range(NUM_FACES)):
                flat.append(ki)
    return np.asarray(flat, dtype=np.int64), starts


subkeep_flat, subkeep_starts = _build_subkeeps()


def best_keep_value(keep_values: np.ndarray) -> np.ndarray:
    """For each of the 252 hands, the max ``keep_values`` over its sub-multiset keeps.

    ``keep_values`` has length ``NUM_KEEPS``; returns a length ``NUM_ROLLS`` array. Every
    hand has at least the empty keep and itself as sub-keeps, so all segments are non-empty.
    """
    gathered = keep_values[subkeep_flat]
    return np.maximum.reduceat(gathered, subkeep_starts)

File: web/src/engine/transitions.ts
/**
 * Reroll-transition helpers derived from the loaded tables. Mirrors `solver/transitions.py`.
 *
 * The canonical ROLLS/KEEPS ordering is imported from the data files (not regenerated), so it
 * cannot drift from Python. This module builds the index maps, the per-hand sub-keep structure
 * (the `np.maximum.reduceat` layout), and the two within-turn reductions.
 */

import { NUM_KEEPS, NUM_ROLLS, type Counts } from "./types.js";
import type { EngineData } from "./data.js";

function keyOf(counts: Counts): string {
  return counts.join(",");
}

function buildIndex(list: Counts[]): Map<string, number> {
  const m = new Map<string, number>();
  for (let i = 0; i < list.length; i++) m.set(keyOf(list[i]), i);
  return m;
}

function isSubMultiset(keep: Counts, roll: Counts): boolean {
  for (let f = 0; f < keep.length; f++) if (keep[f] > roll[f]) return false;
  return true;
}

export class Transitions {
  readonly rollIndex: Map<string, number>;
  readonly keepIndex: Map<string, number>;
  readonly emptyKeep = 0;
  /** Start offset into `subkeepFlat` for each roll (roll-outer, keep-inner order). */
  readonly subkeepStarts: Int32Array;
  /** Flattened keep indices that are sub-multisets of each roll. */
  readonly subkeepFlat: Int32Array;
  /** Initial-throw distribution over the 252 hands (= empty-keep row of T). */
  readonly rollProb: Float64Array;
  private readonly T: Float32Array;

  constructor(data: EngineData) {
    this.T = data.T;
    this.rollIndex = buildIndex(data.rolls);
    this.keepIndex = buildIndex(data.keeps);

    const flat: number[] = [];
    this.subkeepStarts = new Int32Array(NUM_ROLLS);
    for (let ri = 0; ri < NUM_ROLLS; ri++) {
      this.subkeepStarts[ri] = flat.length;
      const roll = data.rolls[ri];
      for (let ki = 0; ki < NUM_KEEPS; ki++) {
        if (isSubMultiset(data.keeps[ki], roll)) flat.push(ki);
      }
    }
    this.subkeepFlat = Int32Array.from(flat);

    this.rollProb = new Float64Array(NUM_ROLLS);
    const base = this.emptyKeep * NUM_ROLLS;
    for (let r = 0; r < NUM_ROLLS; r++) this.rollProb[r] = this.T[base + r];
  }

  /** `T @ e` : (462 x 252) · (252) -> (462). Per keep, expected downstream value. */
  matvecT(e: Float64Array): Float64Array {
    const out = new Float64Array(NUM_KEEPS);
    for (let k = 0; k < NUM_KEEPS; k++) {
      let acc = 0;
      const rowBase = k * NUM_ROLLS;
      for (let r = 0; r < NUM_ROLLS; r++) acc += this.T[rowBase + r] * e[r];
      out[k] = acc;
    }
    return out;
  }

  /**
   * For each of the 252 hands, the max `keepValues` over its sub-multiset keeps.
   * `keepValues` has length NUM_KEEPS; returns length NUM_ROLLS. Every hand has at least the
   * empty keep and itself as sub-keeps, so all segments are non-empty.
   */
  bestKeepValue(keepValues: Float64Array): Float64Array {
    const out = new Float64Array(NUM_ROLLS);
    const flat = this.subkeepFlat;
    for (let ri = 0; ri < NUM_ROLLS; ri++) {
      const start = this.subkeepStarts[ri];
      const end = ri + 1 < NUM_ROLLS ? this.subkeepStarts[ri + 1] : flat.length;
      let m = -Infinity;
      for (let j = start; j < end; j++) {
        const v = keepValues[flat[j]];
        if (v > m) m = v;
      }
      out[ri] = m;
    }
    return out;
  }

  /**
   * Argmax companion to `bestKeepValue`: for each hand, the keep index attaining the segment
   * max of `keepValues`. Strict `>` keeps the first (lowest) keep index on ties, matching the
   * Python `_best_keep_arg` (`solver/findings.py`). Returns length NUM_ROLLS.
   */
  bestKeepArg(keepValues: Float64Array): Int32Array {
    const out = new Int32Array(NUM_ROLLS);
    const flat = this.subkeepFlat;
    for (let ri = 0; ri < NUM_ROLLS; ri++) {
      const start = this.subkeepStarts[ri];
      const end = ri + 1 < NUM_ROLLS ? this.subkeepStarts[ri + 1] : flat.length;
      let m = -Infinity;
      let arg = flat[start];
      for (let j = start; j < end; j++) {
        const v = keepValues[flat[j]];
        if (v > m) {
          m = v;
          arg = flat[j];
        }
      }
      out[ri] = arg;
    }
    return out;
  }

  /** Dot product of two equal-length vectors (used for `rollProb · e1`). */
  dot(a: Float64Array, b: Float64Array): number {
    let acc = 0;
    for (let i = 0; i < a.length; i++) acc += a[i] * b[i];
    return acc;
  }
}
